Link the register jump

qwe661234 opened the pull request:
Originally, the interpreter would exit the execution loop if the instruction was a register jump, which posed a critical performance issue. To address this problem, we now search for the next block at the end of the execution of a register jump. If we find the next block, we link the register jump to it.

As shown in below analysis, this modification improves the overall performance by 1~2%.

* Core i7-11700

| Metric  | Original       |    Proposed    | Speedup |
|---------|----------------|----------------|-------|
| CoreMark| 1931.047 iter/s| 1956.64 iter/s | +1.33%|
| stream  |   79.156 sec   |   77.123 sec   | +2.57%|

Close: #223

jserv commented on src/emulate.c:
```diff
@@ -265,6 +265,65 @@ void rv_debug(riscv_t *rv)
+    block->insn_capacity = 1 << bits;
+    block->n_insn = 0;
+    block->predict = NULL;
+    block->ir = malloc(block->insn_capacity * sizeof(rv_insn_t));
```
Add assertion on the return value of `malloc`.

jserv commented on src/emulate.c:
```diff
@@ -265,6 +265,65 @@ void rv_debug(riscv_t *rv)
+/* try to locate an already translated block in the block map */
+static block_t *block_find(const block_map_t *map, const uint32_t addr)
+{
+    assert(map);
```
Assert on `addr` as well.

qwe661234 replied:
It is possible that the addr is zero, so we can add `assert(addr)` here.

qwe661234 committed:
Link the register jump

Originally, the interpreter would exit the execution loop if the instruction
was a register jump, which posed a critical performance issue. To address
this problem, we now search for the next block at the end of the execution
of a register jump. If we find the next block, we link the register jump
to it.

As shown in below analysis, this modification improves the overall
performance by 1~2%.

* Core i7-11700

| Metric  |    Original    |    Propursed   |Speedup|
|---------+----------------+----------------+-------|
| CoreMark| 1931.047 iter/s| 1956.64 iter/s | +1.33%|
| stream  |   79.156 sec   |   77.123 sec   | +2.57%|

Close: #223
